### DISCUSSION TITLE
How to properly cache Rector cache in Github Actions?

### DISCUSSION THREAD

**Discussion by CanVuralStudocu** [rectorphp/rector#8357] [Q&A]
Hello,

[Documentation](https://getrector.com/documentation/cache-in-ci) recommends this cache key `${{ runner.os }}-rector-${{ hashFiles('**/composer.lock') }}` to be used in Github Actions.

But this cache key only depends on the changes of composer.lock files. Does not change with the actual code file changes. For example if a new file is added or removed from the codebase, Rector cache would not be aware of that.

Does anyone use a different cache key? Like a key created from the all files in a directory or something?

**Comment by ondrejmirtes** [comment]
The only problem with that cache key is that it doesn’t get updated so the cache makes less and less sense over time until it’s invalidated thanks to composer.lock being updated.

You can use a trick in GitHub Actions that:

1) will update the cache every time
2) will find the correct latest cache

the trick is to have a unique key each time, by appending run ID to the end, and then have the common prefix in `restore-keys`.

Like mentioned here (also applies to PHPStan’s result cache): https://staabm.github.io/2023/10/21/phpstan-result-cache-gotchas.html

**Reply by CanVuralStudocu** [reply]
Ooh that makes sense. I didn't understand what `restoreKeys` was doing before. But it makes sense!

I'll try this. Thanks.

**Reply by samsonasik** [reply]
@scollovati you may need to create a directory './var/cache/rector` first on ci process, /cc @staabm this may need to be mentioned in doc?

**Reply by samsonasik** [reply]
@scollovati documentation updated https://getrector.com/documentation/cache-in-ci , could you verify if it works? thank you.
